### ISSUE TITLE
System.Text.Json.JsonSerializer DeserializeAsyncEnumerable works slowly with large json files (10+mb)

### ISSUE THREAD

**Issue by olehrechukh** [dotnet/runtime#100582]
### Discussed in https://github.com/dotnet/runtime/discussions/100581

<div type='discussions-op-text'>

<sup>Originally posted by **olehrechukh** April  3, 2024</sup>
I'm using `HttpClient` to make request, `HttpConnection.ChunkedEncodingReadStream` as input stream from the response, and `JsonSerializer.DeserializeAsyncEnumerable` to asynchronously deserialize-enumerate, but unfortunately, it works very slowly for large json objects.

``` csharp
await using var fileStream = File.OpenRead(filePath);
using var content = new MultipartFormDataContent();
using var streamContent = new StreamContent(fileStream);
streamContent.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data")
{
    Name = "file",
    FileName = Path.GetFileName(filePath)
};
content.Add(streamContent);

var httpRequestMessage = new HttpRequestMessage(HttpMethod.Post, "api/v1/streaming")
{
    Content = content
};

var response = await httpClient.SendAsync(httpRequestMessage, HttpCompletionOption.ResponseHeadersRead,
    cancellationToken);

var responseStream = await response.Content.ReadAsStreamAsync(cancellationToken);
var streamWrapper = new StreamWrapper(responseStream);

var bufferSize = 10 * 1024 * 1024;
var options = new JsonSerializerOptions
{
    PropertyNameCaseInsensitive = true,
    Converters = {new JsonElementToObjectConverter()},
    DefaultBufferSize = bufferSize
};

await foreach (var batchPart in JsonSerializer.DeserializeAsyncEnumerable<BatchPart>(streamWrapper, options,
                   cancellationToken))
{
}
```

After investigation, I discovered that `JsonSerializer` is trying to read `DefaultBufferSize`, but the server/input still does not have enough data to fill it. After this, it tries to deserialize the object, not expecting the buffer to be filled. It works very slowly when taking into account thousands of iterations.
```
Time            Iteration #N            Buffer to read          Read bytes
12:35:45.145            1               16777216                535
12:35:49.327            2               16777216                4090
12:35:49.33             3               16773126                10671
12:35:49.33             4               16762455                4088
12:35:49.331            5               16758367                10673
...
12:36:04.751            882             24051345                4088
12:36:04.789            883             24047257                10658
12:36:04.827            884             24036599                4088
12:36:04.871            885             24032511                10667
...
```

In general, it takes ~15 minutes, but the same case only takes ~2 seconds if I am not using chunked-reading.

``` csharp
var response = await httpClient.SendAsync(httpRequestMessage, HttpCompletionOption.ResponseContentRead,
    cancellationToken);
```

The inner implementation explicitly uses `fillBuffer: false` for deserialization.

https://github.com/dotnet/runtime/blob/3bae822e56590dec674ed25df8e5b7f23927af9a/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Stream.cs#L492

1. Why `fillBuffer` parameter is hard-coded? Should it be configurable from `JsonSerializerOptions`?
2. How I can fix performance issue for large json?</div>

**Comment by olehrechukh** [comment]
There is source code

https://github.com/olehrechukh/SlowJsonDeserialization

**Comment by eiriktsarpalis** [comment]
Hi @olehrechukh, I believe this was a performance issue [that recently got fixed for .NET 9](https://github.com/dotnet/runtime/pull/96856). I tried running your repro using .NET 9 Preview 2 and performance now looks comparable:

```
Count=1
Request-response ResponseHeadersRead takes 480.0093 ms for 15777877 bytes
Request-response ResponseContentRead takes 398.2504 ms for 15777872 bytes
Diff between ResponseHeadersRead and ResponseContentRead response is 81.7589 ms

Count=2
Request-response ResponseHeadersRead takes 563.509 ms for 31555741 bytes
Request-response ResponseContentRead takes 425.4144 ms for 31555747 bytes
Diff between ResponseHeadersRead and ResponseContentRead response is 138.0946 ms

Count=3
Request-response ResponseHeadersRead takes 699.0776 ms for 47333622 bytes
Request-response ResponseContentRead takes 648.9553 ms for 47333619 bytes
Diff between ResponseHeadersRead and ResponseContentRead response is 50.1223 ms

Count=4
Request-response ResponseHeadersRead takes 1072.6692 ms for 63111484 bytes
Request-response ResponseContentRead takes 832.3162 ms for 63111485 bytes
Diff between ResponseHeadersRead and ResponseContentRead response is 240.353 ms

Count=5
Request-response ResponseHeadersRead takes 1149.0854 ms for 78889360 bytes
Request-response ResponseContentRead takes 1037.6158 ms for 78889366 bytes
Diff between ResponseHeadersRead and ResponseContentRead response is 111.4696 ms
```

I can't provide any workarounds for .NET 8 other than the ones you already suggested, unfortunately.

**Comment by olehrechukh** [comment]
Thank you, I hope .NET 9 will resolve the issue.

**Comment by eiriktsarpalis** [comment]
Fixed by #96856.
